### PR TITLE
Tweaks Camera - it can capture revenant/floor cluwne + fixes quirk codes

### PR DESCRIPTION
**Pull request by EvilDragonfiend**
## About The Pull Request
Camera related changes
* Rename quirk file names. `Good.dm`-`Negative.dm` is a weird relation
* Camera Obscura can capture revenant and floor cluwne as well
* added a static list that a camera can see invisible mobs
* Changes photographer quirk to Psychic Photographer 

## Why It's Good For The Game
Haunting creatures that can be detected by a camera obscura can make another story by related with them.
Photographer quirk is too weak and is not almost meaningful. When considering Camera Obscura is easy to make, It's now given to photographer quirk holders at default.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

image

camera obscura

image

normal camera

image

normal camera + ghost wizard event

</details>

## Changelog
:cl:
code: changed weird whitespace indentation into tab indentation in cluwne code
code: renamed quirk file names
code: minor change to cluwne code (mob layer related)
add: added a static list that a camera can see invisible mobs. any object/mob being invisible but in this list will be revealed by camera obscura.
add: camera obscura now can detect Floor Cluwne, Revenant as well.
tweak: Photographer quirk is now Psychic Photographer which gives you a camera obscura.
code: client_alert() proc that alerts players without pausing any code
code: quirk preference will be properly reset if quirks are tweaked in the future.
/:cl:

### CODE REVIEW

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]
Using flags_1 unnecessarilly bloats the scope of this singular item.

flags_1 is good for:
 - Things that will apply to all atoms (INVISIBILITY_REVEALED_BY_CAMERA only ever applies to a small number of types and is only used by 1 item)
 - Things that apply to instances rather than types (Initialized applies when an atom is initialised, INVISIBILITY_REVEALED_BY_CAMERA applies to the types rather than the instances)

There are 3 alternative solutions that I can see being relatively simple:
 - Component (Probably not the best considering DCS has some overhead. Applies to instances rather than types too meaning that for this use-case its less optimal).
 - Have a isCameraVisible proc. (Same problem with flags in that it adds a new proc to everything in the entire game, making an unnecessarilly wide scope)
 - Create a lookup table on the camera which maps types to if they can be seen (the 3 types that can be seen have their type as key and TRUE as a value). I personally think this implementation is the best and the list can be made static to save on memory and init times (There are no camera subtypes which show different kinds of ghosts, and if that is ever added it would be easy to change (Just make it not static)).

**Comment by EvilDragonfiend** [comment]
changed

File: code/modules/photography/camera/camera.dm [hunk 1]
```diff
@@ -39,6 +38,13 @@
 	var/can_customise = TRUE
 	var/default_picture_name
 
+	var/see_ghosts = CAMERA_NO_GHOSTS //for the spoop of it
+	var/static/list/detectable_invisible_atom = list(
+		/mob/dead/observer,
+		/mob/living/simple_animal/revenant,
+		/mob/living/simple_animal/hostile/floor_cluwne
+	)
```
Comment (PowerfulBacon) [inline code/modules/photography/camera/camera.dm]: Use an associative list and use lookups instead of 'in'.
(I think byond's dictionary implementation is good, so going off of regular implementations this should be true)

Using in requires checking every element in the list to see if it is correct. This is O(n) (The time it takes to execute this is proportional to the size of the list).
```
for(var/i in list)
  if (i == thing)
    return i
return null
```

Using a dictionary lookup generates a hash and then uses that to check if the element is in the dictionary. This is O(1) (No matter how many elements are in the list, it takes the same amount of time to check)
```
//The hash is an index of the list in the range 0 to n (n >= number of elements in the list (The more nulls there are in the list, the more memory you waste, but the more optimised this is as you get less hash collisions))
hashed_index = calc_hash(thing)
var/located = list[hashed_index]
//To account for hash collisions (2 things have the same hash)
//Keep looking through the list until we find null, or the thing we are looking for
//If 2 things have the same hash, then it will be inserted in the next free slot
while (located  != null)
  if (located == thing)
    return thing
  hashed_index ++
  located  = list[hashed_index]
//Its not in the list
return null
```

TL;DR dictionaries have better scaling factor, so using a lookup will probably be better here (Although since n is 3 it probably doesn't make a whole lot of difference).

**Review by PowerfulBacon — CHANGES_REQUESTED** [review]

File: code/modules/photography/camera/camera.dm (hunk 1 above)
Comment (PowerfulBacon) [reply code/modules/photography/camera/camera.dm]:
```dm
var/static/list/detectable_invisible_atom = list(
/mob/dead/observer = TRUE,
/mob/living/simple_animal/revenant = TRUE,
/mob/living/simple_animal/hostile/floor_cluwne = TRUE
)

if (detectable_invisible_atom[target.type])
  //We can see the thing!
  see_thing()
```

Comment (EvilDragonfiend) [reply code/modules/photography/camera/camera.dm]: wow giving TRUE for each key like that. interesting.
and yeah, it's only n=3 and it won't go a big number. I think it doesn't have to be hashed.

**Comment by EvilDragonfiend** [comment]
local test done.

**Review by PowerfulBacon — DISMISSED** [review]
Can remove the proc to similfy, but that works. (Note that it doesn't actually return TRUE or FALSE, it returns TRUE or null which may cause some confusion as a proc)

**Comment by EvilDragonfiend** [comment]
then `return detectable_invisible_atom[A.type] || FALSE` might work, I think. I don't think null is fine to be here.

**Comment by PowerfulBacon** [comment]
The if statement can handle nulls, so it can be put directly in if its really necessary

File: code/modules/photography/camera/camera.dm [hunk 2]
```diff
@@ -212,6 +218,8 @@
 	after_picture(user, P, flag)
 	blending = FALSE
 
+/obj/item/camera/proc/checks_detectable(atom/A)
+	return detectable_invisible_atom[A.type] || FALSE
```
Comment (PowerfulBacon) [inline code/modules/photography/camera/camera.dm]:
```suggestion
	return !!detectable_invisible_atom[A.type]
```

Comment (EvilDragonfiend) [reply code/modules/photography/camera/camera.dm]: ah that's cool way to do so.

File: code/datums/traits/positive_quirk.dm [hunk 3]
```diff
@@ -135,16 +135,16 @@
 	eyes.Insert(H) //refresh their eyesight and vision
 
 /datum/quirk/photographer
-	name = "Photographer"
-	desc = "You know how to handle a camera, shortening the delay between each shot."
+	name = "Psychic Photographer"
+	desc = "You have a special camera that can capture a photo of ghosts. Also, your experience to capture ghosts shortened the delay between each shot."
```
Comment (itsmeow) [inline code/datums/traits/positive_quirk.dm]: Your experience in photography shortens the delay

File: code/modules/photography/camera/camera.dm [hunk 4]
```diff
@@ -39,6 +38,13 @@
 	var/can_customise = TRUE
 	var/default_picture_name
 
+	var/see_ghosts = CAMERA_NO_GHOSTS //for the spoop of it
+	var/static/list/detectable_invisible_atom = list(
```
Comment (itsmeow) [inline code/modules/photography/camera/camera.dm]: use a typecache

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by EvilDragonfiend** [comment]
changed and local test done.

File: code/modules/photography/camera/camera.dm [hunk 5]
```diff
@@ -39,6 +38,20 @@
...
+	. = ..()
+	if(!length(detectable_invisible_atom))
+		detectable_invisible_atom = typecacheof(list(
+			// put detactable atom list here
+			/mob/dead/observer,
+			/mob/living/simple_animal/revenant,
+			/mob/living/simple_animal/hostile/floor_cluwne
+		))
+		for(var/each in detectable_invisible_atom)
+			detectable_invisible_atom[each] = TRUE
```
Comment (itsmeow) [inline code/modules/photography/camera/camera.dm]: no need for this. use is_in_typecache

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Comment by EvilDragonfiend** [comment]
changed

File: code/datums/traits/positive_quirk.dm [hunk 6]
```diff
@@ -135,16 +135,16 @@
 	eyes.Insert(H) //refresh their eyesight and vision
 
 /datum/quirk/photographer
-	name = "Photographer"
-	desc = "You know how to handle a camera, shortening the delay between each shot."
+	name = "Psychic Photographer"
```
Comment (itsmeow) [inline code/datums/traits/positive_quirk.dm:138]: how does this affect existing characters with the quirk? Do they keep it?

File: code/modules/photography/camera/camera.dm [hunk 7]
```diff
@@ -191,7 +202,7 @@
 				blueprints = TRUE
 	for(var/mob/M in mobs)
 		// No describing invisible stuff (except ghosts)!
-		if((M.invisibility >= SEE_INVISIBLE_LIVING || M.alpha <= 50) && !isobserver(M))
+		if(M.alpha <= 50 || !(!(M.invisibility >= SEE_INVISIBLE_LIVING) || (see_ghosts && checks_detectable(M))))
```
Comment (itsmeow) [inline code/modules/photography/camera/camera.dm]: This boolean logic is really unreadable

File: code/modules/photography/camera/camera.dm [hunk 8]
```diff
@@ -212,6 +223,8 @@
 	after_picture(user, P, flag)
 	blending = FALSE
 
+/obj/item/camera/proc/checks_detectable(atom/A)
+	return is_type_in_typecache(A, detectable_invisible_atom)
```
Comment (itsmeow) [inline code/modules/photography/camera/camera.dm:227]: this proc is pretty useless now tbh
at minimum rename it to can_camera_see_atom

**Review by itsmeow — CHANGES_REQUESTED** [review]

File: code/datums/traits/positive_quirk.dm (hunk 6 above)
Comment (EvilDragonfiend) [reply code/datums/traits/positive_quirk.dm:138]: Well it looks quirk code is really badly made. It removes but through a weird way. I will fix that too.

File: code/modules/photography/camera/camera.dm (hunk 8 above)
Comment (EvilDragonfiend) [reply code/modules/photography/camera/camera.dm:227]: > this proc is pretty useless now tbh

since the if logic is difficult to read, this is better than using checking_type_cache directly. the proc name is changed now.

**Comment by EvilDragonfiend** [comment]
image

```
/proc/_client_alert(client/C, message, title)
	alert(C, message, title)

/proc/client_alert(client/C, message, title)
	addtimer(CALLBACK(GLOBAL_PROC, .proc/_client_alert, C, message, title), 0)
```

I don't know if we have any proc or helper that works like this. It alerts players without pausing any code.

File: code/datums/traits/positive_quirk.dm (hunk 6 above)
Comment (EvilDragonfiend) [reply code/datums/traits/positive_quirk.dm:138]: quirks will be reset if these are not eligible. check a post below please.

**Comment by llol111** [comment]
i like this idea for floor clowns but im not sure for revenants

**Comment by zeskorion** [comment]
inb4 increased knowledge and availability leads to a camera obscura ghost metagame meta

**Comment by EvilDragonfiend** [comment]
> inb4 increased knowledge and availability leads to a camera obscura ghost metagame meta

okay, but... how?

File: code/controllers/subsystem/processing/quirks.dm [hunk 9]
```diff
@@ -36,14 +39,18 @@ PROCESSING_SUBSYSTEM_DEF(quirks)
...
-			stack_trace("Invalid quirk \"[V]\" in client [cli.ckey] preferences")
-			cli.prefs.active_character.all_quirks -= V
-			badquirk = TRUE
-	if(badquirk)
+			stack_trace("Invalid quirk \"[V]\" in client [cli.ckey] preferences. the game has reset their quirks automatically.")
+			bad_quirk_checker = BAD_QUIRK_FLAG
+	if(bad_quirk_checker) // negative & zero value = calculation good / positive quirk value = something's wrong
+		cli.prefs.active_character.all_quirks = list()
 		cli.prefs.active_character.save(cli)
+		client_alert(cli, "You have an outdated quirk, or more of them. Your eligible quirks are kept at this round, but your character preference has been reset. Please re-apply your quirks.", "Oh, no!")
```
Comment (itsmeow) [inline code/controllers/subsystem/processing/quirks.dm]: Just have it say
`Your selected character quirk \"[V]\" is no longer present in the game. Your selected quirks have been reset, please review them at any time.`

Comment (EvilDragonfiend) [reply code/controllers/subsystem/processing/quirks.dm:54]: The failsafe covers both case of removal or quirk point adjustment rather than handling specific case. Your correction isn't how it works.

**Comment by EvilDragonfiend** [comment]
> Take picture of guy Guy has lots of ghosts He is antag
> 
> Maybe hide ghosts if they are orbiting a mob?

That's even more frequently done by curators and chasing them is usually a bad idea.
Also orbiting isn't properly projected in photo. All ghosts are just captured as if they're standing, not orbiting.

File: code/controllers/subsystem/processing/quirks.dm (hunk 9 above)
Comment (EvilDragonfiend) [reply code/controllers/subsystem/processing/quirks.dm]: This failsafe is to cover both of two cases - quirk removal and quirk point adjustment. your correction is not appliable to quirk point adjustment because you don't know which one was adjusted.
So, this failsafe is made to cover things generally.

Comment (itsmeow) [reply code/controllers/subsystem/processing/quirks.dm]: It should say the quirk name still so they know, also it should say "one or more outdated quirks"

Comment (EvilDragonfiend) [reply code/controllers/subsystem/processing/quirks.dm]: changed

File: code/controllers/subsystem/processing/quirks.dm [hunk 10]
```diff
@@ -40,17 +38,16 @@ PROCESSING_SUBSYSTEM_DEF(quirks)
...
-			bad_quirk_checker = BAD_QUIRK_FLAG
-	if(bad_quirk_checker) // negative & zero value = calculation good / positive quirk value = something's wrong
+			bad_quirks += V
+	if(bad_quirk_checker || length(bad_quirks)) // negative & zero value = calculation good / positive quirk value = something's wrong
 		cli.prefs.active_character.all_quirks = list()
 		cli.prefs.active_character.save(cli)
-		client_alert(cli, "You have an outdated quirk, or more of them. Your eligible quirks are kept at this round, but your character preference has been reset. Please re-apply your quirks.", "Oh, no!")
-
-#undef BAD_QUIRK_FLAG
+		client_alert(cli, "You have one or more outdated quirks[length(bad_quirks) ? ": [english_list(bad_quirks)]" : ""]. Your eligible quirks are kept at this round, but your character preference has been reset. please review them at any time.", "Oh, no!")
```
Comment (itsmeow) [inline code/controllers/subsystem/processing/quirks.dm]: Capitalize Please

**Review by itsmeow — CHANGES_REQUESTED** [review]

**Review by itsmeow — APPROVED** [review]
